File: lib/java/src/main/java/io/github/msgpact/MsgPactSchema.java
package io.github.msgpact;

import static io.github.msgpact.internal.schema.CreateMsgPactSchemaFromFileJsonMap.createMsgPactSchemaFromFileJsonMap;
import static io.github.msgpact.internal.schema.GetSchemaFileMap.getSchemaFileMap;

import java.util.List;
import java.util.Map;

import io.github.msgpact.internal.types.VFieldDeclaration;
import io.github.msgpact.internal.types.VType;

/**
 * A parsed msgPact schema.
 */
public class MsgPactSchema {

    public final List<Object> original;
    public final Map<String, VType> parsed;
    public final Map<String, VFieldDeclaration> parsedRequestHeaders;
    public final Map<String, VFieldDeclaration> parsedResponseHeaders;

    public MsgPactSchema(List<Object> original,
            Map<String, VType> parsed,
            Map<String, VFieldDeclaration> parsedRequestHeaders,
            Map<String, VFieldDeclaration> parsedResponseHeaders) {
        this.original = original;
        this.parsed = parsed;
        this.parsedRequestHeaders = parsedRequestHeaders;
        this.parsedResponseHeaders = parsedResponseHeaders;
    }

    public static MsgPactSchema fromJson(String json) {
        return createMsgPactSchemaFromFileJsonMap(Map.of("auto_", json));
    }

    public static MsgPactSchema fromFileJsonMap(Map<String, String> fileJsonMap) {
        return createMsgPactSchemaFromFileJsonMap(fileJsonMap);
    }

    public static MsgPactSchema fromDirectory(String directory) {
        final var map = getSchemaFileMap(directory);
        return createMsgPactSchemaFromFileJsonMap(map);
    }
}

File: lib/ts/src/MsgPactSchema.ts
import { createMsgPactSchemaFromFileJsonMap } from './internal/schema/CreateMsgPactSchemaFromFileJsonMap';
import { getSchemaFileMap } from './internal/schema/GetSchemaFileMap';
import { VFieldDeclaration } from './internal/types/VFieldDeclaration';
import { VType } from './internal/types/VType';
import { FsModule, PathModule } from './fileSystem';

export class MsgPactSchema {
    /**
     * A parsed msgPact schema.
     */

    original: any[];
    full: any[];
    parsed: Record<string, VType>;
    parsedRequestHeaders: Record<string, VFieldDeclaration>;
    parsedResponseHeaders: Record<string, VFieldDeclaration>;

    constructor(
        original: any[],
        full: any[],
        parsed: Record<string, VType>,
        parsedRequestHeaders: Record<string, VFieldDeclaration>,
        parsedResponseHeaders: Record<string, VFieldDeclaration>,
    ) {
        this.original = original;
        this.full = full;
        this.parsed = parsed;
        this.parsedRequestHeaders = parsedRequestHeaders;
        this.parsedResponseHeaders = parsedResponseHeaders;
    }

    static fromJson(json: string): MsgPactSchema {
        return createMsgPactSchemaFromFileJsonMap({ auto_: json });
    }

    static fromFileJsonMap(fileJsonMap: Record<string, string>): MsgPactSchema {
        return createMsgPactSchemaFromFileJsonMap(fileJsonMap);
    }

    /**
     * @param fs - node fs
     * @param path - node path
     */
    static fromDirectory(directory: string, fs: FsModule, path: PathModule): MsgPactSchema {
        // TODO
        const m = getSchemaFileMap(directory, fs, path);
        return createMsgPactSchemaFromFileJsonMap(m);
    }
}
